src/Npgsql/Replication/NpgsqlReplicationConnection.cs

```diff
@@ -0,0 +1,888 @@
+        {
+            EnsureState(ReplicationConnectionState.Streaming);
+            if (await _feedbackSemaphore.WaitAsync(Timeout.Infinite, cancellationToken))
+                await SendFeedback(cancellationToken: cancellationToken);
+        }
+
+        async Task SendFeedback(bool requestReply = false, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                // Disable the timers while we are sending
+                _sendFeedbackTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+                _requestFeedbackTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+
+                var connector = _npgsqlConnection.Connector!;
+                var buf = connector.WriteBuffer;
+
+                const int len = 39;
+
+                if (buf.WriteSpaceLeft < len)
+                    await connector.Flush(true, cancellationToken);
+
+                buf.WriteByte(FrontendMessageCode.CopyData);
+                buf.WriteInt32(len - 1);
+                buf.WriteByte((byte)'r');  // TODO: enum/const?
+                // We write the LSNs as Int64 here to save us the casting
+                buf.WriteInt64(Interlocked.Read(ref _lastReceivedLsn));
+                buf.WriteInt64(Interlocked.Read(ref _lastFlushedLsn));
+                buf.WriteInt64(Interlocked.Read(ref _lastAppliedLsn));
+                buf.WriteInt64(TimestampHandler.ToPostgresTimestamp(DateTime.Now));
+                buf.WriteByte(requestReply ? (byte)1 : (byte)0);
+
+                await connector.Flush(true, cancellationToken);
+            }
+            finally
+            {
+                // Restart the timers if we are still streaming
+                if (State == ReplicationConnectionState.Streaming)
+                {
+                    _sendFeedbackTimer.Change(WalReceiverStatusInterval, Timeout.InfiniteTimeSpan);
+                    _requestFeedbackTimer.Change(_requestFeedbackInterval, Timeout.InfiniteTimeSpan);
+                }
+                _feedbackSemaphore.Release();
+            }
+        }
+
+        async void TimerRequestFeedback(object? obj)
+        {
+            try
+            {
+                if (State != ReplicationConnectionState.Streaming)
+                    return;
+
+                if (await _feedbackSemaphore.WaitAsync(Timeout.Infinite))
+                    await SendFeedback(true);
+            }
+            catch (Exception e)
+            {
+                Log.Error("An exception occurred while requesting streaming replication feedback from the server.", e, _npgsqlConnection?.Connector?.Id ?? 0);
+            }
+        }
+
+        async void TimerSendFeedback(object? obj)
+        {
+            try
+            {
+                if (State != ReplicationConnectionState.Streaming)
+                    return;
+
+                if (await _feedbackSemaphore.WaitAsync(TimeSpan.Zero))
```

Comment:
As there is no timeout (immediate exit), shouldn't we just use a synchronous alternative?

Reply:
Nope - I took a look at the references. `WaitAsync` should actually be more optimized.